configuration/configurator/SchemaElement.cpp

```diff
@@ -0,0 +1,887 @@
+
+            if (pRef != NULL && *pRef != 0 && pElement->getConstAncestorNode(3)->getNodeType() != XSD_SCHEMA)
+            {
+                pElement->setRef(pRef);
+                CConfigSchemaHelper::getInstance()->addElementForRefProcessing(pElement);
+            }
+            else
+            {
+                // TODO:  throw exception
+            }
+        }
+        else if (strcmp(iterAttrib->queryName(), XML_ATTR_DEFAULT) == 0)
+        {
+            const char *pDefault = iterAttrib->queryValue();
+
+            assert(pDefault != NULL);
+            assert(strlen(pElement->getDefault()) == 0);
+            pElement->setDefault(pDefault);
+        }
+        assert(iterAttrib->queryValue() != NULL);
+    }
+    assert(strlen(pElement->getName()) > 0);
+
+    if (strlen(pElement->getRef()) != 0)
+    {
+        assert(pElement->getComplexTypeArray()->length() == 0);
+        assert(pElement->getSimpleType() == NULL);
+        assert(strlen(pElement->getDefault()) == 0);
+        assert(strlen(pElement->getType()) == 0);
+        /*assert(pElement->getKey() == NULL);
+        assert(pElement->getKeyRef() == NULL);
+        assert(pElement->getUnique() == NULL);*/
+    }
+
+    ::StringBuffer strXPathExt(xpath);
+
+    strXPathExt.append("/").append(XSD_TAG_ANNOTATION);
+    pElement->m_pAnnotation = CAnnotation::load(pElement, pSchemaRoot, strXPathExt.str());
+
+    strXPathExt.clear().append(xpath).append("/").append(XSD_TAG_COMPLEX_TYPE);
+    pElement->m_pComplexTypeArray = CComplexTypeArray::load(pElement, pSchemaRoot, strXPathExt.str());
+
+    if (pElement->m_pAnnotation != NULL && pElement->m_pAnnotation->getAppInfo() != NULL && strlen(pElement->m_pAnnotation->getAppInfo()->getTitle()) > 0)
+    {
+        /****  MUST FIX TO HAVE CORRECT UI TAB LABELS (but getName is expected to return the XPATH name *****/
+        //pElement->setName(pElement->m_pAnnotation->getAppInfo()->getTitle());
+        pElement->setTitle(pElement->m_pAnnotation->getAppInfo()->getTitle());
+    }
+    else
+        pElement->setTitle(pElement->getName());
+
+/*    strXPathExt.clear().append(xpath).append("/").append(XSD_TAG_KEY);
+    pElement->m_pKeyArray = CKeyArray::load(pElement, pSchemaRoot, strXPathExt.str());
+
+    strXPathExt.clear().append(xpath).append("/").append(XSD_TAG_KEYREF);
+    pElement->m_pKeyRefArray = CKeyRefArray::load(pElement, pSchemaRoot, strXPathExt.str());
+*/
+    strXPathExt.clear().append(xpath).append("/").append(XSD_TAG_SIMPLE_TYPE);
+    pElement->m_pSimpleType = CSimpleType::load(pElement, pSchemaRoot, strXPathExt.str());
+
+    SETPARENTNODE(pElement, pParentNode);
+
+    return pElement;
+}
+
+const CElement* CElement::getTopMostElement(const CXSDNodeBase *pNode)
+{
+    if (pNode == NULL)
+        return NULL;
+    else if (pNode->getNodeType() == XSD_ELEMENT)
+    {
+        if (pNode->getParentNodeByType(XSD_ELEMENT) == NULL)
```

Comment:
The line 192 - 194 can be replaced by: if ((pNode->getNodeType() == XSD_ELEMENT) &&  (pNode->getParentNodeByType(XSD_ELEMENT) == NULL))